tests/sentry/integrations/msteams/test_notify_action.py

```diff
@@ -92,13 +95,78 @@ def test_applies_correctly(self, mock_record):
             alert_id=None,
         )
 
+        assert len(mock_record_event.mock_calls) == 2
+        start, end = mock_record_event.mock_calls
+        assert start.args[0] == EventLifecycleOutcome.STARTED
+        assert end.args[0] == EventLifecycleOutcome.SUCCESS
+
     @responses.activate
+    @patch("sentry.integrations.utils.metrics.EventLifecycle.record_event")
+    @mock.patch("sentry.analytics.record")
+    def test_client_error(self, mock_record, mock_record_event):
+        event = self.get_event()
+
+        rule = self.get_rule(
+            data={"team": self.integration.id, "channel": "Naboo", "channel_id": "nb"}
+        )
+
+        notification_uuid = "123e4567-e89b-12d3-a456-426614174000"
+        results = list(rule.after(event=event, notification_uuid=notification_uuid))
+        assert len(results) == 1
+
+        responses.add(
+            method=responses.POST,
+            url="https://smba.trafficmanager.net/amer/v3/conversations/nb/activities",
+            status=500,
+            json={},
+        )
+
+        results[0].callback(event, futures=[])
+        data = orjson.loads(responses.calls[0].request.body)
+
+        assert "attachments" in data
+        attachments = data["attachments"]
+        assert len(attachments) == 1
+
+        # Wish there was a better way to do this, but we
+        # can't pass the title and title link separately
+        # with MS Teams cards.
+        title_card = attachments[0]["content"]["body"][0]
+        title_pattern = r"\[%s\](.*)" % event.title
+        assert re.match(title_pattern, title_card["text"])
+        mock_record.assert_called_with(
+            "alert.sent",
+            provider="msteams",
+            alert_id="",
+            alert_type="issue_alert",
+            organization_id=self.organization.id,
+            project_id=self.project.id,
+            external_id="nb",
+            notification_uuid=notification_uuid,
+        )
+        mock_record.assert_any_call(
+            "integrations.msteams.notification_sent",
+            category="issue_alert",
+            organization_id=self.organization.id,
+            project_id=self.project.id,
+            group_id=event.group_id,
+            notification_uuid=notification_uuid,
+            alert_id=None,
+        )
+
+        assert len(mock_record_event.mock_calls) == 2
+        start, end = mock_record_event.mock_calls
+        assert start.args[0] == EventLifecycleOutcome.STARTED
+        assert end.args[0] == EventLifecycleOutcome.FAILURE
```

Comment:
If most of this logic is copy-pasted, is it worth pulling into a helper method?

Reply:
ur right, let me add something to test utils

Reply:
added a method, i will update all the other places we do this in another pr